File: Python/DesignPattern/ATM_Machine_Demo.py
"""ATM Machine - State Design Pattern"""

from abc import ABC, abstractmethod


class IATMMachineStates(ABC):
    @abstractmethod
    def insert_card(self):
        pass

    @abstractmethod
    def eject_card(self):
        pass

    @abstractmethod
    def with_draw_cash(self):
        pass


class ATMMachine:
    def __init__(self):
        self.no_card_state = NoCardState(self)
        self.card_inserted_state = CardInsertedState(self)
        self.current_state = self.no_card_state

    def get_no_card_state(self):
        return self.no_card_state

    def get_card_inserted_state(self):
        return self.card_inserted_state

    def set_current_state(self, state: IATMMachineStates):
        self.current_state = state

    def insert_card(self):
        self.current_state.insert_card()

    def eject_card(self):
        self.current_state.eject_card()

    def with_draw_cash(self):
        self.current_state.insert_card()


class NoCardState(IATMMachineStates):
    def __init__(self, atm: ATMMachine):
        self.atm = atm

    def insert_card(self):
        print("Card is Inserted")
        self.atm.set_current_state(self.atm.get_card_inserted_state())

    def eject_card(self):
        print("No Card is present please insert card")

    def with_draw_cash(self):
        print("No Card No Cash")


class CardInsertedState(IATMMachineStates):
    def __init__(self, atm: ATMMachine):
        self.atm = atm

    def insert_card(self):
        print("card is already inserted")

    def eject_card(self):
        print("Card is ejected")
        self.atm.set_current_state(self.atm.get_no_card_state())

    def with_draw_cash(self):
        pass


if __name__ == "__main__":
    atm = ATMMachine()
    atm.insert_card()
    atm.eject_card()

File: C++/DesignPattern/ATM_Machine_Demo.cpp
#include <iostream>
#include <memory>
using namespace std;

// ATM Machine - State Design Pattern

class ATMMachine;

struct IATMMachineStates {
    virtual void insertCard() = 0;
    virtual void ejectCard() = 0;
    virtual void withDrawCash() = 0;
    virtual ~IATMMachineStates() = default;
};

class ATMMachine {
    shared_ptr<IATMMachineStates> currentState;
    shared_ptr<IATMMachineStates> noCardState;
    shared_ptr<IATMMachineStates> cardInsertedState;

public:
    ATMMachine();

    shared_ptr<IATMMachineStates> getNoCardState() { return noCardState; }
    shared_ptr<IATMMachineStates> getCardInsertedState() { return cardInsertedState; }

    void setCurrentState(shared_ptr<IATMMachineStates> state) { currentState = move(state); }

    void insertCard();
    void ejectCard();
    void withDrawCash();
};

class NoCardState : public IATMMachineStates {
    ATMMachine* atm;

public:
    explicit NoCardState(ATMMachine* atm) : atm(atm) {}

    void insertCard() override {
        cout << "Card is Inserted\n";
        atm->setCurrentState(atm->getCardInsertedState());
    }

    void ejectCard() override {
        cout << "No Card is present please insert card\n";
    }

    void withDrawCash() override {
        cout << "No Card No Cash\n";
    }
};

class CardInsertedState : public IATMMachineStates {
    ATMMachine* atm;

public:
    explicit CardInsertedState(ATMMachine* atm) : atm(atm) {}

    void insertCard() override {
        cout << "card is already inserted\n";
    }

    void ejectCard() override {
        cout << "Card is ejected\n";
        atm->setCurrentState(atm->getNoCardState());
    }

    void withDrawCash() override {}
};

ATMMachine::ATMMachine() {
    noCardState = make_shared<NoCardState>(this);
    cardInsertedState = make_shared<CardInsertedState>(this);
    currentState = noCardState;
}

void ATMMachine::insertCard() { currentState->insertCard(); }
void ATMMachine::ejectCard() { currentState->ejectCard(); }
void ATMMachine::withDrawCash() { currentState->insertCard(); }

int main() {
    ATMMachine atm;
    atm.insertCard();
    atm.ejectCard();
    return 0;
}
